README.md
# nz-bank-account-validator

[![npm](https://img.shields.io/npm/v/@fnzc/nz-bank-account-validator.svg)](https://www.npmjs.com/package/@fnzc/nz-bank-account-validator)

_A validator for New Zealand bank accounts_

## Installation

```
npm install --save @fnzc/nz-bank-account-validator
```
or
```
yarn add @fnzc/nz-bank-account-validator
```

## Usage

Function requires four string parameters:
* Bank ID
* Bank Branch
* Account Base
* Suffix
Returns a boolean.

```
var bankAccountValidator = require('@fnzc/nz-bank-account-validator');
bankAccountValidator.isValidNZBankNumber('00', '0000', '00000000', '000');
```

## Algorithm

The algorithm is based on the 'Non-Resident Withholding Tax And Resident Withholding Tax Specification Document' issues by the Inland Revenue on the 31st of March 2016.

### Bank account number validation

The bank account number format used by all banks is numeric and includes the following parts:

* Bank ID (maximum 2 digits)
* Bank branch (maximum 4 digits)
* Account base number (maximum 8 digits)
* Account suffix (maximum 4 digits).
    
For processing at Inland Revenue the fields for the individual account number parts are all of maximum size. If less than the maximum number of digits is supplied, then values are right justified and the fields padded with zeroes (where applicable).

The first step in the validation process is to verify the bank branch number. For every bank ID, a range of branch numbers is allocated. If the bank ID is not one of those listed below or the branch number is not included in the range(s) specified, the bank account number is invalid. If the branch number is valid, then derive the appropriate code from the “Algorithm” column below and perform the second validation step as outlined in the first paragraph over the page.

| Bank ID | Valid Branch Numbers | Algorithm |
|---|---|---|
|01|0001 - 0999, 1100 - 1199, 1800 - 1899|See note|
|02|0001 - 0999, 1200 - 1299|See note|
|03|0001 - 0999, 1300 - 1399, 1500 - 1599, 1700 – 1799 , 1900 - 1999|See note|
|04|2020 - 2024||
|06|0001 - 0999, 1400 - 1499|See note|
|08|6500 - 6599|D|
|09|0000|E|
|10|5165 - 5169|See note|
|11|5000 - 6499, 6600 - 8999|See note|
|12|3000 - 3299, 3400 – 3499, 3600 - 3699|See note|
|13|4900 - 4999|See note|
|14|4700 - 4799|See note|
|15|3900 - 3999|See note|
|16|4400 - 4499|See note|
|17|3300 - 3399|See note|
|18|3500 - 3599|See note|
|19|4600 - 4649|See note|
|20|4100 - 4199|See note|
|21|4800 - 4899|See note|
|22|4000 - 4049|See note|
|23|3700 - 3799|See note|
|24|4300 - 4349|See note|
|25|2500 - 2599|F|
|26|2600 - 2699|G|
|27|3800 - 3849|See note|
|28|2100 - 2149|G|
|29|2150 - 2299|G|
|30|2900 - 2949|See note|
|31|2800 - 2849|X|
|33|6700 - 6799|F|
|35|2400 - 2499|See note|
|38|9000 - 9499|See note|
|88|8800 - 8805||

**Note**: If the account base number is below 00990000 then apply algorithm A, otherwise apply algorithm B.

The second validation step is a modulus n algorithm applied to the whole account number. The algorithm type is derived from the table on the previous page. Follow this process:

* Identify the corresponding weight factor for every digit in the account number as showninthetablebelow. Note:allfields(iebankID,bankbranch,accountbase and account suffix) are right justified and padded with zeroes.
* Add together the products of the weight factors and their associated account number digit. If the algorithm E or G is used then add the two digits of the product (tens and ones), and again the two digits of the result before summing (see example 3 on page 16).
* Divide the sum by the value in the “Modulo” column below. If the remainder is zero then the bank account number is valid.

| Algorithm | Bank | Branch | Account Base | Suffix | Modulo |
|---|---|---|---|---|---|
|A|00|6379|00A58421|0000|11|
|B|00|0000|00A58421|0000|11|
|C|37|0000|91A53421|0000|11|
|D|00|0000|07654321|0000|11|
|E|00|0000|00005432|0001|11|
|F|00|0000|01731731|0000|10|
|G|00|0000|01371371|0371|10|
|X|00|0000|00000000|0000|1|

**Note 1**: Algorithm C is not currently used by the banks.
**Note 2**: Algorithm X (for Bank ID 31) always verifies the bank account number to be valid. It is included in this table so the same validation logic can be applied to all account numbers.

## [Changelog](https://github.com/FNZC/nz-bank-account-validator/blob/master/CHANGELOG.md)

## Contributing

Pull requests are welcome.

## [License](https://github.com/FNZC/nz-bank-account-validator/blob/master/LICENSE)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [1.0.3] - 2019-08-09
### Modified
- Update bankData to match 2020 Resident Withholding Tax (RWT) and Non-Resident Withholding Tax (NRWT).
- Cross reference with https://www.paymentsnz.co.nz/resources/industry-registers/bank-branch-register/


## [1.0.2] - 2019-01-30
### Modified
- Switch from string.padLeft to left-pad library as the first one is crashing in Android RN


## [1.0.1] - 2019-01-30
### Added
- Types file.
### Changed
- Updated the README


## [1.0.0] - 2019-01-29
### Added
- First version of the package.


spec/spec.js
const validator = require('../index.js')

describe("Valid New Zealand Bank Accounts", function() {

  it("should return true for Bank Account number 01090200068389000", function() {
    expect(validator.isValidNZBankNumber('01', '0902', '00068389', '000')).toBe(true);
  });

  it("should return true for Bank Account number 086523019545120001", function() {
    expect(validator.isValidNZBankNumber('08', '6523', '01954512', '0001')).toBe(true);
  });

  it("should return true for Bank Account number 262600003208710032", function() {
    expect(validator.isValidNZBankNumber('26', '2600', '00320871', '0032')).toBe(true);
  });

  it("should return false for Bank Account number 26485749385729888", function() {
    expect(validator.isValidNZBankNumber('26', '4857', '49385729', '888')).toBe(false);
  });

});

index.js
const data = require('./bankData.js')
const leftPad = require('left-pad')
const bankData = data.bankData

const isValidBankAndBranch = (bankId, branch) => {
  if (!bankData.map(o => o.id).includes(+bankId)) return false
  const bank = bankData.find(o => o.id === +bankId)
  return bank.branches.map(b => +branch >= b.from && +branch <= b.to).some(r => r)
}

const getWeightFactor = (bankId, accountNumber) => {
  switch (+bankId) {
    case 8:
      return '000000076543210000'
    case 9:
      return '000000000054320001'
    case 25:
    case 33:
      return '000000017317310000'
    case 26:
    case 28:
    case 29:
      return '000000013713710371'
    case 31:
      return '000000000000000000'
    default:
      if (+accountNumber < 990000) return '00637900A584210000'
      else return '00000000A584210000'
  }
}

const getModulo = (bankId) => {
  switch (+bankId) {
    case 25:
    case 33:
    case 26:
    case 28:
    case 29:
      return 10
    case 31:
      return 1
    default:
      return 11
  }
}

const isValidNZBankNumber = (bk, brch, acct, suf) => {
  const bank = leftPad(bk, 2, '0')
  const branch = leftPad(brch, 4, '0')
  const account = leftPad(acct, 8, '0')
  const suffix = leftPad(suf, 4, '0')
  if (+account === 0) return false
  if (!isValidBankAndBranch(bank, branch)) return false
  const checkString = bank + branch + account + suffix
  if (checkString.length !== 18) return false
  const weightFactor = getWeightFactor(bank, account)
  const result = weightFactor.split('')
    .map((v, i) => (v === 'A' ? 10 : +v) * +checkString
      .split('')[i])
    .reduce((acc, curr) => acc + curr)
  const checkNumber = getModulo(bank)
  return result % checkNumber === 0
}

exports.isValidNZBankNumber = isValidNZBankNumber


bankData.js
const bankData = [
  {
    id: 1,
    branches: [
      {
        from: 1,
        to: 999
      },
      {
        from: 1100,
        to: 1199
      },
      {
        from: 1800,
        to: 1899
      }
    ]
  },
  {
    id: 2,
    branches: [
      {
        from: 1,
        to: 999
      },
      {
        from: 1200,
        to: 1299
      }
    ]
  },
  {
    id: 3,
    branches: [
      {
        from: 1,
        to: 999
      },
      {
        from: 1300,
        to: 1399
      },
      {
        from: 1500,
        to: 1599
      },
      {
        from: 1700,
        to: 1799
      },
      {
        from: 1900,
        to: 1999
      }
    ]
  },
  {
    id: 4,
    branches: [
      {
        from: 2020,
        to: 2024
      }
    ]
  },
  {
    id: 6,
    branches: [
      {
        from: 1,
        to: 999
      },
      {
        from: 1400,
        to: 1499
      }
    ]
  },
  {
    id: 8,
    branches: [
      {
        from: 6500,
        to: 6599
      }
    ]
  },
  {
    id: 9,
    branches: [
      {
        from: 0,
        to: 0
      }
    ]
  },
  {
    id: 10,
    branches: [
      {
        from: 5165,
        to: 5169
      },
    ]
  },
  {
    id: 11,
    branches: [
      {
        from: 5000,
        to: 6499
      },
      {
        from: 6600,
        to: 8999
      }
    ]
  },
  {
    id: 12,
    branches: [
      {
        from: 3000,
        to: 3299
      },
      {
        from: 3400,
        to: 3499
      },
      {
        from: 3600,
        to: 3699
      }
    ]
  },
  {
    id: 13,
    branches: [
      {
        from: 4900,
        to: 4999
      }
    ]
  },
  {
    id: 14,
    branches: [
      {
        from: 4700,
        to: 4799
      }
    ]
  },
  {
    id: 15,
    branches: [
      {
        from: 3900,
        to: 3999
      }
    ]
  },
  {
    id: 16,
    branches: [
      {
        from: 4400,
        to: 4499
      }
    ]
  },
  {
    id: 17,
    branches: [
      {
        from: 3300,
        to: 3399
      }
    ]
  },
  {
    id: 18,
    branches: [
      {
        from: 3500,
        to: 3599
      }
    ]
  },
  {
    id: 19,
    branches: [
      {
        from: 4600,
        to: 4649
      }
    ]
  },
  {
    id: 20,
    branches: [
      {
        from: 4100,
        to: 4199
      }
    ]
  },
  {
    id: 21,
    branches: [
      {
        from: 4800,
        to: 4899
      }
    ]
  },
  {
    id: 22,
    branches: [
      {
        from: 4000,
        to: 4049
      }
    ]
  },
  {
    id: 23,
    branches: [
      {
        from: 3700,
        to: 3799
      }
    ]
  },
  {
    id: 24,
    branches: [
      {
        from: 4300,
        to: 4349
      }
    ]
  },
  {
    id: 25,
    branches: [
      {
        from: 2500,
        to: 2599
      }
    ]
  },
  {
    id: 26,
    branches: [
      {
        from: 2600,
        to: 2699
      }
    ]
  },
  {
    id: 27,
    branches: [
      {
        from: 3800,
        to: 3849
      }
    ]
  },
  {
    id: 28,
    branches: [
      {
        from: 2100,
        to: 2149
      }
    ]
  },
  {
    id: 29,
    branches: [
      {
        from: 2150,
        to: 2299
      }
    ]
  },
  {
    id: 30,
    branches: [
      {
        from: 2900,
        to: 2949
      }
    ]
  },
  {
    id: 31,
    branches: [
      {
        from: 2800,
        to: 2849
      }
    ]
  },
  {
    id: 33,
    branches: [
      {
        from: 6700,
        to: 6799
      }
    ]
  },
  {
    id: 35,
    branches: [
      {
        from: 2400,
        to: 2499
      }
    ]
  },
  {
    id: 38,
    branches: [
      {
        from: 9000,
        to: 9499
      }
    ]
  },
  {
    id: 88,
    branches: [
      {
        from: 8800,
        to: 8805
      }
    ]
  }
]

exports.bankData = bankData
